Cython/Utility/CythonFunction.c

```diff
@@ -75,6 +75,22 @@ typedef struct {
 #define __Pyx_IsCyOrPyCFunction(obj)  __Pyx_TypeCheck2(obj, __pyx_CyFunctionType, &PyCFunction_Type)
 #define __Pyx_CyFunction_CheckExact(obj)  __Pyx_IS_TYPE(obj, __pyx_CyFunctionType)
 
+// Redefine PyCFunction access functions to also support CyFunction.
+// We need our own copies because the inline functions in CPython have a type-check assert
+// that breaks with a CyFunction in debug mode.
+#undef __Pyx_PyCFunction_Check
+#define __Pyx_PyCFunction_Check(func)  __Pyx_IsCyOrPyCFunction(func)
```

Comment:
I'm a little worried about this in the limited API. i.e. will it cause CyFunctions to get passed to functions expecting a `PyCFunction` (bearing in mind that the limited API `CyFunction` doesn't even share the `PyCFunction` layout).

It probably doesn't matter because I suspect most of these are used for optimizations that are skipped in the limited API. But it feels like it could go wrong

Reply:
I agree. I actually wanted to go the other way and make CyFunction inherit from `PyCMethod_Type` in the Limited API in order to make the function usages safe, but then noticed that this actually cannot work at all because the object struct isn't exposed.

I now think it would be better to make the usages explicit, i.e. allow CyFunction _and_ PyCFunction explicitly in the code, and then map the "either or" check function to a plain `PyCFunction_Check()` if CyFunction is _not_ available. That's safer than implicitly allowing both. I'll change that in the PR when I get to it.

Reply:
I hope you like https://github.com/cython/cython/pull/5739/commits/f1aef480e91501035b29a8f88715b35c018b829a better.

Reply:
> I actually wanted to go the other way and make CyFunction inherit from PyCMethod_Type in the Limited API in order to make the function usages safe, but then noticed that this actually cannot work at all because the object struct isn't exposed.

Yes - I did the refactoring to hold a pointer reluctantly for the Limited API but couldn't see much else that was likely to work.

> I hope you like https://github.com/cython/cython/commit/f1aef480e91501035b29a8f88715b35c018b829a better.

Yes - generally looks more explicit about what it's actually doing which I think is better.